### PR TITLE
[Move selected_rows PR #2] Added Selected_Rows and rw_lock to Pten

### PR DESCRIPTION
**Pull request by veyron95**
### PR types
Others

### PR changes
Others 

### Describe
A series of jobs where selected_rows moves from Fluid to Pten.
There are 5 PR to complete the migration work in expected.

PR 1: Renamed selected_rows.* to selected_rows_utils.* [PR39037](https://github.com/PaddlePaddle/Paddle/pull/39037)

#### This is PR 2: Added selected_rows.* and rw_lock.h to Pten

 **NOTE:**
"paddle/fluid/framework/mixed_vector.h" will be removed soon by @Aurelius84（留杰）.
"paddle/fluid/framework/data_type.h" will be removed in PR4.

#### follow-up work：
PR 3: Change the relationship of [include/Cmake].
PR 4: VisitDataType -> Pten::DataType.
PR 5: SelectedRows inherits from TensorBase.

### CODE REVIEW

**Review by YuanRisheng — DISMISSED** [review]

**Review by YuanRisheng — DISMISSED** [review]

File: paddle/pten/tests/core/test_rw_lock.cc [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#include "paddle/pten/core/rw_lock.h"
+
+#include <gtest/gtest.h>  // NOLINT
+#include <thread>         // NOLINT
+
+void f1(pten::RWLock *lock) {
```
Comment (chenwhql) [inline paddle/pten/tests/core/test_rw_lock.cc:23]: pten下的test实现希望均在pten::tests命名空间下，之前fluid的不太规范，希望能加一下，可以参考其他单测

File: paddle/pten/tests/core/test_selected_rows.cc [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#include <time.h>
+#include <thread>  // NOLINT
+
+#include "gtest/gtest.h"
+#include "paddle/pten/core/selected_rows.h"
+
+namespace pten {
```
Comment (chenwhql) [inline paddle/pten/tests/core/test_selected_rows.cc:21]: 同上

File: paddle/pten/core/rw_lock.h [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+#include <mutex>  // NOLINT
+#endif            // !_WIN32
+
+// See Note [ Why still include the fluid headers? ]
+#include "paddle/fluid/platform/enforce.h"
+
+namespace pten {
+
+#if !defined(_WIN32)
+struct RWLock {
```
Comment (chenwhql) [inline paddle/pten/core/rw_lock.h]: 之后，pten/core下根目录的头文件会暴露给外部用户做外部开发，这个rw_lock.h从定位上，是一个特定场景下使用的组件，希望能放到pten/core/utils下，命名空间上也可以加一层pten::utils，外部用户不会用到这个头文件；此外，原来fluid的rw_lock.h后后续PR会移除吧?

Comment (veyron95) [reply paddle/pten/core/rw_lock.h]: 好的，修改一下。fluid的rw_lock.h 在后续PR会移除。

File: paddle/pten/tests/core/test_rw_lock.cc (hunk 1 above)
Comment (veyron95) [reply paddle/pten/tests/core/test_rw_lock.cc:23]: Done, thx.

File: paddle/pten/tests/core/test_selected_rows.cc (hunk 2 above)
Comment (veyron95) [reply paddle/pten/tests/core/test_selected_rows.cc:21]: Done, thx.

**Review by chenwhql — APPROVED** [review]

**Review by XieYunshen — APPROVED** [review]

**Pull request merged**